renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp

```diff
@@ -943,9 +943,22 @@ VkResult WrappedVulkan::vkQueueSubmit(VkQueue queue, uint32_t submitCount,
         GetResourceManager()->MarkResourceFrameReferenced(GetResID(fence), eFrameRef_Read);
 
       rdcarray<VkResourceRecord *> maps;
+      rdcarray<VkResourceRecord *> lockedRecords;
       {
         SCOPED_LOCK(m_CoherentMapsLock);
         maps = m_CoherentMaps;
+        // For memory that is potentially going to be flushed, lock the memory map state
+        // so that it is not unmapped.
+        for(auto it = maps.begin(); it != maps.end(); ++it)
+        {
+          VkResourceRecord *record = *it;
+          MemMapState &state = *record->memMapState;
+          if(state.mapCoherent)
+          {
+            state.mrLock.Lock();
```

Comment:
I think we should be able to make the locking more finer grained than this, if I understand the problem correctly. We only care about one memory map racing with itself, so it doesn't matter if we're flushing memory A while memory B gets unmapped.

So I think instead we can lock inside the loop below before checking the `MemMapState` contents - that way we ensure the state isn't modified while we're looking at it. If we're processing memory A and memory B gets unmapped, then when we get to memory B we lock and check and find it's no longer map so we can skip it.